packages/next/build/webpack/plugins/flight-types-plugin.ts

```diff
@@ -0,0 +1,133 @@
+  fetchCache?: 'auto' | 'force-no-store' | 'only-no-store' | 'default-no-store' | 'default-cache' | 'only-cache' | 'force-cache'
+  preferredRegion?: 'auto' | 'home' | 'edge'
+}
+
+// =============
+// Utility types
+type RevalidateRange<T> = T extends { revalidate: any } ? NonNegative<T['revalidate']> : never
+type Impossible<K extends keyof any> = { [P in K]: never }
+function check<Base, T extends Base>(_mod: T & Impossible<Exclude<keyof T, keyof Base>>): void {}
+
+// https://github.com/sindresorhus/type-fest
+type Numeric = number | bigint
+type Zero = 0 | 0n
+type Negative<T extends Numeric> = T extends Zero ? never : \`\${T}\` extends \`-\${string}\` ? T : never
+type NonNegative<T extends Numeric> = T extends Zero ? T : Negative<T> extends never ? T : '__invalid_negative_number__'
+`
+}
+
+export class FlightTypesPlugin {
+  appDir: string
+  dev: boolean
+  isEdgeServer: boolean
+
+  constructor(options: Options) {
+    this.appDir = options.appDir
+    this.dev = options.dev
+    this.isEdgeServer = options.isEdgeServer
+  }
+
+  apply(compiler: webpack.Compiler) {
+    const assetPrefix = this.dev
+      ? '..'
+      : this.isEdgeServer
+      ? '..'
+      : path.join('..', '..')
+
+    const handleModule = (_mod: webpack.Module, assets: any) => {
+      if (_mod.layer !== WEBPACK_LAYERS.server) return
+      const mod: webpack.NormalModule = _mod as any
+
+      if (!mod.resource) return
+      if (!mod.resource.startsWith(this.appDir + path.sep)) return
+      if (!/\.(js|jsx|ts|tsx|mjs)$/.test(mod.resource)) return
+
+      const IS_LAYOUT = /[/\\]layout\.[^.]+$/.test(mod.resource)
+      const IS_PAGE = !IS_LAYOUT && /[/\\]page\.[^.]+$/.test(mod.resource)
+      const relativePath = path.relative(this.appDir, mod.resource)
+
+      // const RSC = mod.buildInfo.rsc
+
+      const typePath = path.join(
+        'types',
+        'app',
+        relativePath.replace(/\.(js|jsx|ts|tsx|mjs)$/, '.ts')
+      )
+      const relativeImportPath = path.join(
+        path.relative(typePath, ''),
+        'app',
+        relativePath.replace(/\.(js|jsx|ts|tsx|mjs)$/, '')
+      )
+      const assetPath = path.join(assetPrefix, typePath)
+
+      if (IS_LAYOUT) {
+        assets[assetPath] = new sources.RawSource(
+          createTypeGuardFile(mod.resource, relativeImportPath, {
+            type: 'layout',
+          })
+        ) as unknown as webpack.sources.RawSource
+      } else if (IS_PAGE) {
+        assets[assetPath] = new sources.RawSource(
+          createTypeGuardFile(mod.resource, relativeImportPath, {
+            type: 'page',
+          })
+        ) as unknown as webpack.sources.RawSource
+      }
+    }
+
+    compiler.hooks.make.tap(PLUGIN_NAME, (compilation) => {
+      compilation.hooks.processAssets.tap(
+        {
+          name: PLUGIN_NAME,
+          stage: webpack.Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE_HASH,
+        },
+        (assets) => {
+          for (const mod of compilation.modules) {
```

Comment:
Do you really need to visit all modules here? Can't you find the modules from the entry points.